File: C++/315.cpp
__________________________________________________________________________________________________
sample 20 ms submission
template <typename T> inline T safe_cast(size_t size) {
  assert(size <= static_cast<size_t>((std::numeric_limits<T>::max)()));
  return static_cast<T>(size);
}

static const int _ = [](){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 0;
}();

class Solution {
public:
  
// A binary index tree with value type T and index type U.
template <typename T, typename U = long long> class BinaryIndexTree {
  static_assert(std::numeric_limits<U>::is_integer);

public:
  // Create a BIT for range [1, Size] with initial value Val.
  BinaryIndexTree(U Size, T Val) : Size(Size), RangeSums(Size, Val) {}

  // \return Range sum of [1, Idx].
  int get(U Idx) {
    U Sum = 0;
    while (Idx > 0) {
      Sum += RangeSums[Idx - 1];
      Idx -= Idx & (~Idx + 1);
    }
    return Sum;
  }

  // Increase value at Idx by one and update all super ranges.
  void update(U Idx) {
    while (Idx - 1 < Size) {
      RangeSums[Idx - 1]++;
      Idx += Idx & (~Idx + 1);
    }
  }

private:
  // Size of the range covered by this BIT.
  U Size;

  // The sum of elements in range [x - lowerbit(x) + 1, x].
  std::vector<T> RangeSums;
};

std::vector<int> countSmaller(std::vector<int> &Nums) {
  // Handle corner cases.
  const auto N = safe_cast<int>(Nums.size());
  if (N < 1)
    return {};

  // Shift the values to [2, inf] since binary index tree must operate on
  // index >= 1 and we would query [1, Nums[i] - 1].
  std::vector<long long> ShiftNums;
  long long MinNum = *std::min_element(Nums.begin(), Nums.end());
  long long MaxNum = std::numeric_limits<long long>::min();
  for (long long Num : Nums) {
    ShiftNums.push_back(Num - MinNum + 2);
    MaxNum = std::max(MaxNum, ShiftNums.back());
  }

  // Hold the number of elements in each range [1, i <= MaxNum].
  BinaryIndexTree<int> BIT(MaxNum, 0);

  // Counts[i] = number of elements in range [1, ShiftNums[i] - 1].
  std::vector<int> SmallerCnts(N, 0);
  for (int i = N - 1; i >= 0; --i) {
    SmallerCnts[i] = BIT.get(ShiftNums[i] - 1);
    BIT.update(ShiftNums[i]);
  }
  return SmallerCnts;
}
  
};
__________________________________________________________________________________________________
sample 9656 kb submission
class Solution {
public:
    vector<int> countSmaller(vector<int>& nums) {
        int n = nums.size();
        vector<int> ret(n);
        for ( int i=0; i<n ;i++ ) {
            int k = 0;
            for ( int j=i+1; j<n; j++ ) {
                if ( nums[j] < nums[i] ) k++;
            }
            ret[i] = k;
        }
        return ret;
    }
};
__________________________________________________________________________________________________

File: Java/315.java
__________________________________________________________________________________________________
sample 3 ms submission
class Solution {
  
  class Node{
    int val;
    int self;
    int child;
    
    Node left;
    Node right;
    
    public Node(int val) {
      this.val = val;
      this.child = 1;
    }
  }
  
  Integer[] ans;
  
  public List<Integer> countSmaller(int[] nums) {
    ans = new Integer[nums.length];
    
    if (nums == null) return null;
    if (nums.length == 0) return new ArrayList<Integer>();
    
    Node head = new Node(nums[nums.length-1]);
    ans[nums.length -1] = 0;
    
    for (int i = nums.length - 2; i >= 0; i--) {
      add(head, i, nums[i], 0); 
    }
    
    return Arrays.asList(ans);
  }
  
  private void add(Node head, int i, int val, int acc) {
    // if i > head.val, go right, add acc to head's val and pass it down
    // if i < head.val go left, increase head's acc by 1
    // if i == head.val, update the value, increase self
    if (val > head.val) {
      // go right
      if (head.right == null) {
        Node right = new Node(val);
        head.right = right;
        ans[i] = acc + head.child + head.self;
      } else {
        add(head.right, i, val, acc + head.child + head.self); 
      }
    } else if (val < head.val) {
      // go left
      head.child++;
      if (head.left == null) {
        Node left = new Node(val);
        head.left = left;
        ans[i] = acc;
      } else {
        add(head.left, i, val, acc);
      }
    } else {
      ans[i] = acc + head.child - 1;
      head.self++;
    }
  }
}
__________________________________________________________________________________________________
sample 36220 kb submission
class Solution {
    
    int[] bit;
    int[] result;
    int n;
    int maxNumber = 100001;
    void update(int idx) {
        while(idx<=maxNumber) {
            bit[idx]+=1;
            idx += idx & -idx;
        }
    }
    
    int find(int idx) {
        int sum = 0;
        while(idx > 0) {
            sum+=bit[idx];
            idx-=idx & -idx;
        }
        return sum;
    }
    
    public List<Integer> countSmaller(int[] nums) {
        n=nums.length;
        if (n==0) {
            return new ArrayList<>();
        }
        if (n==1) {
            return Arrays.asList(0);
        }
        int min = Integer.MAX_VALUE;
        for (int i=n-1;i>=0;i--) {
            min = Math.min(min, nums[i]);
        }
        if (min <= 0) {
            min--;
            for (int i=n-1;i>=0;i--) {
                nums[i] -= min;
            }
        }
        bit = new int[maxNumber];    
        result = new int[n];
        for (int i=n-1;i>=0;i--) {
            result[i] = find(nums[i]-1);
            update(nums[i]);
        }
        List<Integer> res = new ArrayList<>();
        for (int i=0;i<n;i++) {
            res.add(result[i]);
        }
        return res;
    }
}
__________________________________________________________________________________________________
